### DISCUSSION TITLE
Cancel a zip operation

### DISCUSSION THREAD

**Discussion by NansySalem** [101arrowz/fflate#165] [Q&A]
Hello,

I'm using fflate's async zip to zip a couple of files. The function then triggers a download. I'm trying to add a functionality to cancel the the operation if the user presses a cancel button. I'm trying to use the return of the zip function, i.e., AyncTerminable, and then in the canceling function terminate the process. I'm getting an error that terminate doesn't exist on AsyncTerminable. I was wondering if someone can shed some lights on how this can be achieved to stop the zipping process as the documentation is really limited in that regards.

Thanks

**Comment by 101arrowz** [comment]
The value returned by `zip` is actually a function, not an object with a `terminate` method. You'd use it like this:

```js
import { strToU8, zip } from 'fflate';

const terminateZip = zip({ 'myfile.txt': strToU8('hello') }, (err, res) => {
  // use zip buffer here
});

cancelButton.onclick = () => {
  terminateZip();
}
```

To use this with `async/await` and promises:

```js
let terminateZip;
const zipPromise = new Promise((resolve, reject) => {
  terminateZip = zip({ 'myfile.txt': strToU8('hello') }, (err, res) => err ? reject(err) : resolve(res));
});

// Now you can await zipPromise for the file, and call terminateZip whenever you want to kill the operation
// Note that terminateZip will cause the promise to never resolve or reject
```

**Comment by NansySalem** [comment]
Thanks for clarifying. Ok, so now if I want to test that the zip thread has
been terminated using jest, what would be the best way to do that? I have a
dialog that on dismiss should terminate the zip process if it has started.

**Reply by 101arrowz** [reply]
I'd just spy on the terminate function if possible. There's no way to tell if the process is dead or not beyond checking if the terminate function was called.

**Comment by NansySalem** [comment]
I have another question: so if we have a zipping process going on and we
call the terminating function, will the terminating function throw or give
any signal that the termination has been done?

**Reply by 101arrowz** [reply]
The termination is immediate and never fails, so you can assume that the process is dead the line after you call the terminate function.
